external/java/src/main/java/ch/epfl/dedis/calypso/Encryption.java

```diff
@@ -1,78 +1,90 @@
 package ch.epfl.dedis.calypso;
 
 import ch.epfl.dedis.lib.exception.CothorityCryptoException;
-import com.google.protobuf.ByteString;
 
-import javax.crypto.BadPaddingException;
-import javax.crypto.Cipher;
-import javax.crypto.IllegalBlockSizeException;
-import javax.crypto.NoSuchPaddingException;
-import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.*;
+import javax.crypto.spec.GCMParameterSpec;
 import javax.crypto.spec.SecretKeySpec;
 import java.security.InvalidAlgorithmParameterException;
 import java.security.InvalidKeyException;
 import java.security.NoSuchAlgorithmException;
 import java.security.SecureRandom;
 
 public class Encryption {
-    public static String algo = "AES/CBC/PKCS5Padding";
-    public static String algoKey = "AES";
-    public static int ivLength = 16;
+    public static final String ALGO = "AES/GCM/NoPadding";
+    public static final String ALGO_KEY = "AES";
+    public static final int KEY_LENGTH = 16;
+    public static final int IV_LENGTH = 12; // standard IV length for GCM
+    public static final int GCM_TLEN = 128;
 
-    public static class keyIv{
-        public byte[] symmetricKey;
-        public byte[] iv;
-        public IvParameterSpec ivSpec;
-        public SecretKeySpec keySpec;
+    /**
+     * KeyIV represents a secret key and an IV.
+     */
+    public static class KeyIv {
+        final byte[] symmetricKey;
+        final byte[] iv;
+        final GCMParameterSpec gcmSpec;
+        final SecretKeySpec keySpec;
 
-        public keyIv(byte[] keyMaterial) throws CothorityCryptoException{
-            int symmetricLength = keyMaterial.length - ivLength;
-            if (symmetricLength <= 0){
-                throw new CothorityCryptoException("too short symmetricKey material");
+        /**
+         * Construct KeyIV from some keyMaterial.
```

Comment:
Need to remind them that "For a given key, the IV must never repeat." Thus, encourage them to use the 0-argument constructor to get a new ephemeral key with random IV.

Or maybe: do not allow them to pass in IV at all, have this constructor take the key material and choose the IV.

Reply:
I updated the comment. I'm not changing the way how IV is passed in because that constructor also needs to be used when decrypting the secret. We have a constructor for `Document` which uses the default constructor to create the key material.